From 0f2fca2b561bd15b8f415181f9f55bc470bd3da9 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:03 -0400
Subject: [scsi] qla2xxx: Add logic to abort BSG commands for ISPFX00

Message-id: <1408391817-6389-16-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87937
O-Subject: [PATCH RHEL 7.1 e-stor 15/69] qla2xxx: Add logic to abort BSG commands for ISPFX00.
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Armen Baloyan <armen.baloyan@qlogic.com>

scsi commit id 767157c5d362152938d22b803944f6a3163ed94f

Signed-off-by: Armen Baloyan <armen.baloyan@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_mr.c b/drivers/scsi/qla2xxx/qla_mr.c
index 5755dea..5c44dd0 100644
--- a/drivers/scsi/qla2xxx/qla_mr.c
+++ b/drivers/scsi/qla2xxx/qla_mr.c
@@ -1883,6 +1883,7 @@ qlafx00_fx_disc(scsi_qla_host_t *vha, fc_port_t *fcport, uint16_t fx_type)
    goto done_free_sp;
   }
   break;
+ case FXDISC_ABORT_IOCTL:
  default:
   break;
  }
@@ -2011,7 +2012,11 @@ qlafx00_fx_disc(scsi_qla_host_t *vha, fc_port_t *fcport, uint16_t fx_type)
   ql_dump_buffer(ql_dbg_init + ql_dbg_buffer, vha, 0x0146,
       (uint8_t *)pinfo, 16);
   memcpy(vha->hw->gid_list, pinfo, QLAFX00_TGT_NODE_LIST_SIZE);
- }
+ } else if (fx_type == FXDISC_ABORT_IOCTL)
+  fdisc->u.fxiocb.result =
+      (fdisc->u.fxiocb.result == cpu_to_le32(0x68)) ?
+      cpu_to_le32(QLA_SUCCESS) : cpu_to_le32(QLA_FUNCTION_FAILED);
+
  rval = le32_to_cpu(fdisc->u.fxiocb.result);
 
 done_unmap_dma:
@@ -2110,6 +2115,10 @@ qlafx00_abort_command(srb_t *sp)
   /* Command not found. */
   return QLA_FUNCTION_FAILED;
  }
+ if (sp->type == SRB_FXIOCB_DCMD)
+  return qlafx00_fx_disc(vha, &vha->hw->mr.fcport,
+      FXDISC_ABORT_IOCTL);
+
  return qlafx00_async_abt_cmd(sp);
 }
 
diff --git a/drivers/scsi/qla2xxx/qla_mr.h b/drivers/scsi/qla2xxx/qla_mr.h
index a5ddea6..c8b8acd 100644
--- a/drivers/scsi/qla2xxx/qla_mr.h
+++ b/drivers/scsi/qla2xxx/qla_mr.h
@@ -343,6 +343,7 @@ struct config_info_data {
 #define FXDISC_GET_TGT_NODE_INFO 0x80
 #define FXDISC_GET_TGT_NODE_LIST 0x81
 #define FXDISC_REG_HOST_INFO  0x99
+#define FXDISC_ABORT_IOCTL  0xff
 
 #define QLAFX00_HBA_ICNTRL_REG  0x20B08
 #define QLAFX00_ICR_ENB_MASK            0x80000000
-- 
1.7.1